### PR TITLE
[MemCpyOpt] No need to create `memcpy(a <- a)`

### PR DESCRIPTION
**Pull request by dianqk**
When forwarding `memcpy`, we don't need to create `memcpy(a, a)`.

I fixed a related test case at https://github.com/llvm/llvm-project/commit/62b3e68d33035f14e83320b7921c80b36dcbea8c.

### CODE REVIEW

File: llvm/test/Transforms/MemCpyOpt/memcpy.ll [hunk 1]
```diff
@@ -139,6 +139,56 @@ define void @test6_memcpy(ptr %src, ptr %dest) nounwind {
...
+}
+
+; There is data writing between the two memcpy operations.
+define void @test6_memcpy_forward_back_failed(ptr %arg) nounwind {
+; CHECK-LABEL: @test6_memcpy_forward_back_failed(
+; CHECK-NEXT:    [[TMP:%.*]] = alloca [16 x i8], align 1
+; CHECK-NEXT:    [[SRC:%.*]] = getelementptr inbounds i8, ptr [[ARG:%.*]], i64 1
+; CHECK-NEXT:    [[DEST:%.*]] = getelementptr inbounds i8, ptr [[ARG]], i64 1
+; CHECK-NEXT:    call void @llvm.memcpy.inline.p0.p0.i32(ptr align 1 [[TMP]], ptr align 1 [[SRC]], i32 16, i1 false)
+; CHECK-NEXT:    call void @llvm.memcpy.inline.p0.p0.i32(ptr align 1 [[DEST]], ptr align 1 [[TMP]], i32 16, i1 false)
```
Comment (dtcxzyw) [inline llvm/test/Transforms/MemCpyOpt/memcpy.ll:181]: `test6_memcpy_forward_back_failed` is optimized into `test6_memcpy_forward_back`. Is it correct?

File: llvm/test/Transforms/MemCpyOpt/memcpy.ll [hunk 2]
```diff
@@ -139,6 +139,56 @@ define void @test6_memcpy(ptr %src, ptr %dest) nounwind {
   ret void
 }
 
+; FIXME: When forwarding to memcpy(arg+1, arg+1), we don't need to create this memcpy.
```
Comment (dtcxzyw) [inline llvm/test/Transforms/MemCpyOpt/memcpy.ll:142]: Remove this FIXME.

File: llvm/test/Transforms/MemCpyOpt/memcpy.ll (hunk 1 above)
Comment (DianQK) [reply llvm/test/Transforms/MemCpyOpt/memcpy.ll]: It seems that this test case is redundant.

File: llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp [hunk 3]
```diff
@@ -1161,6 +1161,14 @@ bool MemCpyOptPass::processMemCpyMemCpyDependence(MemCpyInst *M,
                      MSSA->getMemoryAccess(MDep), MSSA->getMemoryAccess(M)))
     return false;
 
+  // No need to create `memcpy(a <- a)`.
+  if (BAA.isMustAlias(M->getDest(), MDep->getRawSource())) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:1165]:
```suggestion
  if (!M->isVolatile() && BAA.isMustAlias(M->getDest(), MDep->getRawSource())) {
```
> The optimizers must not change the number of volatile operations or change their order of execution relative to other volatile operations.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp (hunk 3 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:1165]: It's already checked here: https://github.com/llvm/llvm-project/blob/385ace339a9fc6fcff429d30da5667a25f0a9888/llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp#L1670

Comment (nikic) [inline llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp:1165]: There is an odd mix of getDest and getRawSource here. I'd expect either both or neither to use getRaw...

Comment (DianQK) [reply llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp]: I changed to `getDest` and `getSource`.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**